### ISSUE TITLE
BUG: np.float64("inf").astype(np.int32) is negative on x86 but positive on ppc64le

### ISSUE THREAD

**Issue by sandipmgiri** [numpy/numpy#9040]
The following test passes on x86, but fails the second assert on ppc64le:
```python
import numpy as np
from numpy.testing import assert_equal

intmin = np.int32(-2147483648)
inf = np.float64(np.inf)
assert_equal(np.int32(-inf), intmin)  # ok on both platforms
assert_equal(np.int32(inf), intmin)   # fails on ppc43le: 2147483647 != -2147483648
```

---

I ran the below sample code in a standalone python file on x86 as well as on ppc64le:

```python
import numpy as np

#### FOR np.inf
print("\n****************************** np.inf **************************************\n")
a = np.inf
x = np.array([a])
print("data type of np.array([inf])===>", np.array([a]).dtype)
y = x.astype(np.int32)
print("int32 range for positive np.inf=====> ", y)

#### FOR -np.inf
print("\n****************************** -np.inf **************************************\n")
b = -np.inf
p = np.array([b])
print("data type of np.array([-inf])===>", np.array([b]).dtype)
q = p.astype(np.int32)
print("int32 range for negative np.inf=====> ", q)
print("\n")
```

On X86  I am getting same output in both the cases (getting min int32 range) ====>
`array([-2147483648], dtype=int32))`

However on ppc64le getting different results  :
for `np.inf`  (getting max int32 range) ====>    `array([2147483647], dtype=int32))`
for -`np.inf` (getting min int32 range) ====>    `array([-2147483648], dtype=int32))`

Looks like the bug in numpy on ppc64le. 
Actually I just wanted the same results on ppc64le as well (like x86). For that what we have to do ? 

I went through the files `/numpy/numpy/core/src/multiarray/multiarraymodule.c` and `/numpy/numpy/core/src/multiarray/methods.c` to find out the reason but unable to find/understand the corresponding code.

If I have to implement the changes for ppc64le, which could be the right place to do so?

**Comment by eric-wieser** [comment]
So less verbosely, the issue is what `np.int32(np.inf)` and `np.int32(-np.inf)` mean?

**Comment by sandipmgiri** [comment]
Actually in our test case we are comparing the lower and upper datatype range . 
On `x86` we are getting lower int32 range for `np.inf` and `-np.inf` , so we have written a test case to compare with lower int32 range. However this test is failing on `ppc64le` , getting max and min range.

**Comment by eric-wieser** [comment]
Can you confirm that this test passes on x86 for you, but fails on ppc64le?
```
from numpy.testing import assert_equal

intmin = np.int32(-2147483648)
assert_equal(np.int32(np.float64(-np.inf)), intmin)
assert_equal(np.int32(np.float64(np.inf)), intmin)
```

**Comment by sandipmgiri** [comment]
Yes , this test is passing on X86, but failing on ppc64le with the following error :

```
bash-4.2# python test.py
Traceback (most recent call last):
  File "test.py", line 6, in <module>
    assert_equal(np.int32(np.float64(np.inf)), intmin)
  File "/usr/lib64/python2.7/site-packages/numpy/testing/utils.py", line 425, in assert_equal
    raise AssertionError(msg)
AssertionError:
Items are not equal:
 ACTUAL: 2147483647
 DESIRED: -2147483648
```

**Comment by eric-wieser** [comment]
Mind editing your issue with that much shorter example? Right now, it's kinda hard to follow

**Comment by sandipmgiri** [comment]
Sure.

**Comment by eric-wieser** [comment]
Note that doing this cast [incurs undefined behaviour](https://stackoverflow.com/questions/38795544/is-casting-of-infinity-to-integer-undefined), so it's hard to say which platform is incorrect here.

I'd argue that we should just issue a warning and leave the platform-specific behaviour>

**Comment by eric-wieser** [comment]
`numpy\core\src\multiarray\lowlevel_strided_loops.c.src` is the place to look, inside `@prefix@_cast_@name1@_to_@name2@`

**Comment by sowani** [comment]
@eric-wieser I checked the discussion at the "cast incurs undefined behavior" link and it seems fair. But I have a query regarding this. Let's assume we just issue a warning from numpy and keep the platform-specific behavior as is. Also, assume that a certain package tries to cast "np.inf" to int32. Now this code will give a certain result on x86 platform. But when one tries to use this package on ppc64le, the result will be incorrect. So there are a couple of obvious solutions which I can think of:
1. Fixing that particular package to handle different expected cast values for x86 and ppc64le
2. Fix the cast operation itself, so that the package need not know about this issue

If we issue a warning from numpy, that might be helpful provided the package owner attempts to build/use the package on ppc64le. However it will still be a bug for all other packages which try to do such cast. So would it be more effective to fix this issue at the numpy level? I would like to understand comments/suggestions on this. Thanks!

**Comment by eric-wieser** [comment]
> But when one tries to use this package on ppc64le, the result will be incorrect.

This is also true of any C program that tries to cast float to int. If the author of the package included a C extension that does this cast, they still have the problem.

The same argument you use for "we should fix this is numpy, not in the user package" can be applied to "we should fix this in C, not in numpy". Presumably for speed reasons, that argument was rejected during the design of C.

Having said that, we already diverge from C behaviour in some place (eg, integer promotion for arithmetic and comparison), so it wouldn't be unreasonable to add special casing here.

**Comment by sandipmgiri** [comment]
Added ppc64le specific condition in TF test case and now its passing. 
Resolved in Tensorflow PR, relevant link -  https://github.com/tensorflow/tensorflow/pull/10522  (merged commit https://github.com/tensorflow/tensorflow/commit/d0d2308389307b5bb3db8cff41b9eaf6d0ae67dd)
